From ae59347fa34406d680eab95643a90860c4db7c83 Mon Sep 17 00:00:00 2001
From: Carlos Calderon <carlos@emuex.com>
Date: Thu, 6 Sep 2018 09:02:03 +0100
Subject: [PATCH] AAEONC: Update for APL4 on watchdog f71808e driver

---
 drivers/watchdog/f71808e_wdt.c | 16 +++++++++++-----
 1 file changed, 11 insertions(+), 5 deletions(-)

diff --git a/drivers/watchdog/f71808e_wdt.c b/drivers/watchdog/f71808e_wdt.c
index c69c394..66d157b 100644
--- a/drivers/watchdog/f71808e_wdt.c
+++ b/drivers/watchdog/f71808e_wdt.c
@@ -70,6 +70,7 @@
 #define SIO_F71889_ID		0x0723	/* Chipset ID */
 #define SIO_F81865_ID		0x0704	/* Chipset ID */
 #define SIO_F81866_ID		0x1010	/* Chipset ID */
+#define SIO_F81804_ID		0x1502	/* Chipset ID */
 
 #define F71808FG_REG_WDO_CONF		0xf0
 #define F71808FG_REG_WDT_CONF		0xf5
@@ -136,7 +137,7 @@ MODULE_PARM_DESC(start_withtimeout, "Start watchdog timer on module load with"
 	" given initial timeout. Zero (default) disables this feature.");
 
 enum chips { f71808fg, f71858fg, f71862fg, f71868, f71869, f71882fg, f71889fg,
-	     f81865, f81866};
+	     f81865, f81866, f81804 };
 
 static const char *f71808e_names[] = {
 	"f71808fg",
@@ -148,6 +149,7 @@ static const char *f71808e_names[] = {
 	"f71889fg",
 	"f81865",
 	"f81866",
+	"f81804",
 };
 
 /* Super-I/O Function prototypes */
@@ -401,14 +403,15 @@ static int watchdog_start(void)
 		break;
 
 	case f81866:
-		/* Set pin 70 to WDTRST# */
+	case f81804:
+		/* Set pin 70/29 on F81866/F81804 to WDTRST# */
 		superio_clear_bit(watchdog.sioaddr, SIO_F81866_REG_PORT_SEL,
 				  BIT(3) | BIT(0));
 		superio_set_bit(watchdog.sioaddr, SIO_F81866_REG_PORT_SEL,
 				BIT(2));
 		/*
 		 * GPIO1 Control Register when 27h BIT3:2 = 01 & BIT0 = 0.
-		 * The PIN 70(GPIO15/WDTRST) is controlled by 2Ch:
+		 * The PIN 70/29(GPIO15/WDTRST) is controlled by 2Ch:
 		 *     BIT5: 0 -> WDTRST#
 		 *           1 -> GPIO15
 		 */
@@ -428,7 +431,7 @@ static int watchdog_start(void)
 	superio_select(watchdog.sioaddr, SIO_F71808FG_LD_WDT);
 	superio_set_bit(watchdog.sioaddr, SIO_REG_ENABLE, 0);
 
-	if (watchdog.type == f81865 || watchdog.type == f81866)
+	if (watchdog.type == f81865 || watchdog.type == f81866 || watchdog.type == f81804)
 		superio_set_bit(watchdog.sioaddr, F81865_REG_WDO_CONF,
 				F81865_FLAG_WDOUT_EN);
 	else
@@ -840,6 +843,9 @@ static int __init f71808e_find(int sioaddr)
 	case SIO_F81866_ID:
 		watchdog.type = f81866;
 		break;
+	case SIO_F81804_ID:
+		watchdog.type = f81804;
+		break;
 	default:
 		pr_info("Unrecognized Fintek device: %04x\n",
 			(unsigned int)devid);
@@ -887,4 +893,4 @@ MODULE_AUTHOR("Nick Chen <nick1chen@aaeon.com.tw>");
 MODULE_LICENSE("GPL");
 
 module_init(f71808e_init);
-module_exit(f71808e_exit);
+	module_exit(f71808e_exit);
